source/compiler/qsc_frontend/src/resolve/imports.rs

```diff
@@ -0,0 +1,524 @@
+            .get_mut_or_default(current_namespace)
+            .contains_key(&name.name)
+        {
+            return Err(Error::Duplicate(
+                name.name.to_string(),
+                resolver.globals.format_namespace_name(current_namespace),
+                name.span,
+            ));
+        }
+    }
+
+    let scope = resolver.current_scope_mut();
+    match (
+        scope.importables.entry(name.name.clone()),
+        current_namespace,
+    ) {
+        (Entry::Vacant(entry), _) => {
+            entry.insert(Res::Importable(imported_item.clone()));
+        }
+        (Entry::Occupied(mut entry), None) => {
+            // allow shadowing in non-namespace (block) scopes
+            entry.insert(Res::Importable(imported_item.clone()));
+        }
+        (Entry::Occupied(_), Some(namespace_id)) => {
+            // collision within the namespace scope
+            return Err(Error::Duplicate(
+                name.name.to_string(),
+                resolver.globals.format_namespace_name(namespace_id),
+                name.span,
+            ));
+        }
+    }
+
+    match *imported_item {
+        Importable::Callable(imported_item_id, status) => {
+            scope
+                .terms
+                .insert(name.name.clone(), Res::Item(imported_item_id, status));
+        }
+        Importable::Ty(imported_item_id, status) => {
+            scope
+                .terms
+                .insert(name.name.clone(), Res::Item(imported_item_id, status));
+            scope
+                .tys
+                .insert(name.name.clone(), Res::Item(imported_item_id, status));
+        }
+        Importable::Namespace(namespace_id, _) => {
+            // A direct import of a namespace is the same as an open with an alias
+            resolver.bind_open(
+                valid_item.path,
+                namespace_id,
+                Some(valid_item.name().name.clone()),
+            );
+        }
+    }
+    Ok(())
+}
+
+/// Iterative resolution wrapper that handles the common pattern of
+/// running resolution attempts in a loop until convergence or maximum iterations.
+///
+/// The iterative approach allows imports and exports to refer to each other
+/// regardless of the order that they are declared in the source code.
+///
+/// Consider this example:
+/// ```qsharp
+/// namespace A {
+///     function B() : Unit {}
+/// }
+///
+/// namespace D {
+///     export C.B;  // Depends on C.B being available
+/// }
+///
+/// namespace C {
+///     export A.B;  // Makes A.B available as C.B
+/// }
+/// ```
+///
+/// Here, `D.B` depends on `C.B`, which depends on `A.B`.
+/// The iterative approach resolves these dependencies in multiple passes.
+#[must_use]
+fn iterate_until_done<F>(mut resolver_fn: F) -> Vec<Error>
+where
+    F: FnMut(&mut FxHashMap<NodeId, Result<(), Error>>) -> bool,
+{
+    let mut errors = Vec::new();
+    let mut attempted_items = FxHashMap::default();
+
+    for i in 1..=100 {
```

Comment:
Should probably have this declared as a const at the top of the file. I think that is the pattern we use for other best effect limiting passes in other code.